sim/priest/shadow/spriest.go

```diff
@@ -0,0 +1,223 @@
+package shadow
+
+import (
+	"time"
+
+	"github.com/wowsims/tbc/sim/core"
+	"github.com/wowsims/tbc/sim/core/proto"
+	"github.com/wowsims/tbc/sim/core/stats"
+	"github.com/wowsims/tbc/sim/priest"
+)
+
+func RegisterShadowPriest() {
+	core.RegisterAgentFactory(proto.PlayerOptions_ShadowPriest{}, func(character core.Character, options proto.PlayerOptions, isr proto.IndividualSimRequest) core.Agent {
+		return NewShadowPriest(character, options, isr)
+	})
+}
+
+var ShadowWeavingDebuffID = core.NewDebuffID()
+var ShadowWeaverAuraID = core.NewAuraID()
+
+func NewShadowPriest(character core.Character, options proto.PlayerOptions, isr proto.IndividualSimRequest) *ShadowPriest {
+	shadowOptions := options.GetShadowPriest()
+
+	// Only undead can do Dev Plague
+	if shadowOptions.Rotation.UseDevPlague == true && options.Race != proto.Race_RaceUndead {
+		shadowOptions.Rotation.UseDevPlague = false
+	}
+
+	selfBuffs := priest.SelfBuffs{}
+
+	basePriest := priest.NewPriest(character, selfBuffs, *shadowOptions.Talents)
+	spriest := &ShadowPriest{
+		Priest:   basePriest,
+		rotation: *shadowOptions.Rotation,
+	}
+
+	if basePriest.Talents.ShadowWeaving > 0 {
+		const dur = time.Second * 15
+		const misDur = time.Second * 24
+
+		swAura := core.Aura{
+			ID:   ShadowWeavingDebuffID,
+			Name: "Shadow Weaving",
+			OnBeforeSpellHit: func(sim *core.Simulation, spellCast *core.SpellCast, spellEffect *core.SpellEffect) {
+				spellEffect.DamageMultiplier *= 1 + 0.02*spriest.swStacks
+			},
+			OnPeriodicDamage: func(sim *core.Simulation, spellCast *core.SpellCast, spellEffect *core.SpellEffect, tickDamage float64) float64 {
+				return tickDamage * (1 + 0.02*spriest.swStacks)
+			},
+			OnExpire: func(sim *core.Simulation) {
+				spriest.swStacks = 0
+			},
+		}
+
+		addShadowWeaving := func(sim *core.Simulation, spellEffect *core.SpellEffect) {
+			if spriest.swStacks < 5 {
+				spriest.swStacks++
+				if sim.Log != nil {
+					sim.Log("(%d) Shadow Weaving: stacks on target %0.0f\n", spriest.ID, spriest.swStacks)
+				}
+			}
+			// Just keep replacing it with new expire time.
+			swAura.Expires = sim.CurrentTime + dur
+			spellEffect.Target.ReplaceAura(sim, swAura)
+		}
+		// This is a combined aura for all spriest major on hit effects.
+		//  Shadow Weaving, Vampiric Touch, and Misery
+		spriest.Character.AddPermanentAura(func(sim *core.Simulation) core.Aura {
```

Comment:
Added a bunch of talents together into a single aura that:
1. Applies shadow weaving stacks debuff to target
2. Applies Misery debuff to target
3. VT mana regen (just checks if VT is ticking right now)
4. Darkness/Shadowform bonus damage on dot tick